src/Networking/NexusMods.Networking.HttpDownloader/AdvancedHttpDownloader.cs

```diff
@@ -0,0 +1,734 @@
+        {
+            return state.Chunks.Where(chunk => (chunk.Read < chunk.Size));
+        }
+
+        private async Task DownloaderTask(DownloadState state, ChannelWriter<WriteOrder> writes, CancellationToken cancel)
+        {
+            if (state.TotalSize < 0)
+            {
+                // if we don't know the total size that means we've never received a response from the server so we don't know
+                // if chunking is even possible. For now only download the first chunk on the primary job until we know more
+                var firstChunk = UnfinishedChunks(state).First();
+                await DownloadChunk(state, firstChunk, writes, cancel);
+            }
+
+            int finishReward = 1024;
+
+            // start one task per unfinished chunk. We never start additional tasks but a task may take on another chunks
+            await Task.WhenAll(UnfinishedChunks(state).Select(async chunk =>
+            {
+                using var chunkJob = await _limiter.Begin($"Download Chunk @${chunk.Offset}", Size.From(chunk.Size), cancel);
+
+                try
+                {
+                    while (chunk.Read < chunk.Size)
+                    {
+                        await DownloadChunk(state, chunk, writes, MakeChunkCancelToken(chunk, cancel));
+
+                        // boost source priority on completion, first one gets the biggest boost
+                        chunk.Source!.Priority -= finishReward;
+                        finishReward /= 2;
+
+                        chunk = PickNextChunk(state, chunk);
+                    }
+                }
+                catch (Exception e) when (e is TaskCanceledException or OperationCanceledException)
+                {
+                    // ignore cancelation
+                }
+            }).ToArray());
+        }
+
+        private ChunkState PickNextChunk(DownloadState state, ChunkState chunk)
+        {
+            ChunkState? slowChunk = FindSlowChunk(UnfinishedChunks(state), chunk.Source!, chunk.BytesPerSecond);
+            if (slowChunk != null)
+            {
+                _logger.LogInformation("canceling chunk {}-{} @ {}, downloading at {} kb/s",
+                    slowChunk.Offset, slowChunk.Offset + slowChunk.Size, slowChunk.Source,
+                    slowChunk.KBytesPerSecond);
+
+                chunk = StealWork(slowChunk);
+                lock (state)
+                {
+                    state.Chunks.Add(chunk);
+                }
+            }
+
+            return chunk;
+        }
+
+        private bool IsDownloadSlow(ChunkState chunk, double referenceSpeed)
+        {
+            var now = DateTime.Now;
+            if ((now - chunk.Started).TotalMilliseconds < MinCancelAge)
+            {
+                // don't cancel downloads that were only just started
+                return false;
+            }
+            if ((chunk.Read / chunk.Size) > 0.8)
```

Comment:
Here `chunk.Read` and `chunk.Size` are both integer types (longs); thus cannot produce fractional output after division.  
This will always return either 1 [post completion] or 0.

You'll need to cast to a float here.